In [5]:
import numpy as np

def print_matrix(matrix, label):
    print(label)
    for row in matrix:
        print(row)

def generate_elementary_matrix(n, i, j, factor):
    E = [[1 if p == q else 0 for q in range(n)] for p in range(n)]
    E[j][i] = factor
    return E

def multiply_matrices(A, B):
    result = [[0 for _ in range(len(B[0]))] for _ in range(len(A))]
    for i in range(len(A)):
        for j in range(len(B[0])):
            for k in range(len(B)):
                result[i][j] += A[i][k] * B[k][j]
    return result

def lu_decomposition(A):
    n = len(A)
    L = [[0 for _ in range(n)] for _ in range(n)]
    U = [row.copy() for row in A]

    for i in range(n):
        for j in range(i + 1, n):
            factor = U[j][i] / U[i][i]
            L[j][i] = factor
            # Applying elementary operation to U
            E = generate_elementary_matrix(n, i, j, -factor)
            print("Elementry Matrix ",E)            
            U = multiply_matrices(E, U)

    for i in range(n):
        L[i][i] = 1

    return L, U

def verify_lu_decomposition(A, L, U):
    reconstructed_A = multiply_matrices(L, U)
    for i in range(len(A)):
        for j in range(len(A[0])):
            if abs(A[i][j] - reconstructed_A[i][j]) > 1e-8:
                return False
    return True

# Example usage with random input
# np.random.seed(42)  # Set seed for reproducibility
n = 3
A = np.random.randint(1, 10, size=(n, n)).tolist()

print_matrix(A, "Original Matrix A:")

# Perform LU decomposition
L, U = lu_decomposition(A)

print_matrix(L, "Lower Triangular Matrix L:")
print_matrix(U, "Upper Triangular Matrix U:")

# Verify A = LU
verification_result = verify_lu_decomposition(A, L, U)
print("Verification Result (A = LU):", verification_result)

Original Matrix A:
[2, 4, 7]
[8, 3, 1]
[4, 2, 8]
Elementry Matrix  [[1, 0, 0], [-4.0, 1, 0], [0, 0, 1]]
Elementry Matrix  [[1, 0, 0], [0, 1, 0], [-2.0, 0, 1]]
Elementry Matrix  [[1, 0, 0], [0, 1, 0], [0, -0.46153846153846156, 1]]
Lower Triangular Matrix L:
[1, 0, 0]
[4.0, 1, 0]
[2.0, 0.46153846153846156, 1]
Upper Triangular Matrix U:
[2.0, 4.0, 7.0]
[0.0, -13.0, -27.0]
[0.0, 0.0, 6.461538461538462]
Verification Result (A = LU): True


<h1> Anshuman </h1>

Original Matrix A:
[2, 4, 7]
[8, 3, 1]
[4, 2, 8]
<br>
Elementry Matrix  [[1, 0, 0], [-4.0, 1, 0], [0, 0, 1]]
Elementry Matrix  [[1, 0, 0], [0, 1, 0], [-2.0, 0, 1]]
Elementry Matrix  [[1, 0, 0], [0, 1, 0], [0, -0.46153846153846156, 1]]
<br>
Lower Triangular Matrix L:
[1, 0, 0]
[4.0, 1, 0]
[2.0, 0.46153846153846156, 1]
<br>
Upper Triangular Matrix U:
[2.0, 4.0, 7.0]
[0.0, -13.0, -27.0]
[0.0, 0.0, 6.461538461538462]
<br>
Verification Result (A = LU): True

<h1> Vishal bhai </h1>
Original Matrix A:
[5, 9, 7]
[2, 4, 9]
[2, 9, 5]
<br>
Elementry Matrix  [[1, 0, 0], [-0.4, 1, 0], [0, 0, 1]]
Elementry Matrix  [[1, 0, 0], [0, 1, 0], [-0.4, 0, 1]]
Elementry Matrix  [[1, 0, 0], [0, 1, 0], [0, -13.500000000000004, 1]]
<br>
Lower Triangular Matrix L:
[1, 0, 0]
[0.4, 1, 0]
[0.4, 13.500000000000004, 1]
<br>
Upper Triangular Matrix U:
[5.0, 9.0, 7.0]
[0.0, 0.3999999999999999, 6.199999999999999]
[0.0, 0.0, -81.50000000000001]
<br>
Verification Result (A = LU): True

<h1> Amit bhai </h1>
Original Matrix A:
[7, 4, 8]
[5, 7, 3]
[7, 8, 5]
<br>
Elementry Matrix  [[1, 0, 0], [-0.7142857142857143, 1, 0], [0, 0, 1]]
Elementry Matrix  [[1, 0, 0], [0, 1, 0], [-1.0, 0, 1]]
Elementry Matrix  [[1, 0, 0], [0, 1, 0], [0, -0.9655172413793105, 1]]
<br>
Lower Triangular Matrix L:
[1, 0, 0]
[0.7142857142857143, 1, 0]
[1.0, 0.9655172413793105, 1]
<br>
Upper Triangular Matrix U:
[7.0, 4.0, 8.0]
[0.0, 4.142857142857142, -2.7142857142857144]
[0.0, 0.0, -0.37931034482758585]
<br>
Verification Result (A = LU): True